Products/Zing/model_updates.py

```diff
@@ -0,0 +1,133 @@
+
+
+import traceback
+
+from itertools import chain
+from logging import getLogger
+
+from zope.component.factory import Factory
+from zope.interface import implements
+
+from Products.ZenModel.Device import Device
+from Products.ZenModel.DeviceComponent import DeviceComponent
+from Products.Zing import fact as ZFact
+from Products.Zing.interfaces import IZingObjectUpdateHandler
+from Products.Zing.tx_state import ZingTxStateManager
+
+
+log = getLogger("zen.zing.model_updates")
+
+
+class ZingObjectUpdateHandler(object):
+    implements(IZingObjectUpdateHandler)
+
+    def __init__(self, context):
+        self.context = context.getDmd()
+        self.zing_tx_state_manager = ZingTxStateManager()
+
+    def is_object_relevant(self, obj):
+        # devices and components with an uuid are relevant
+        uuid = None
+        try:
+            uuid = obj.getUUID()
+        except Exception:
+            pass
+        return uuid and (isinstance(obj, Device) or isinstance(obj, DeviceComponent))
+
+    def _get_zing_tx_state(self):
+        """ """
+        return self.zing_tx_state_manager.get_zing_tx_state(self.context)
+
+    def _update_object(self, obj, idxs=None):
+        if self.is_object_relevant(obj) and isinstance(obj, Device):
+            tx_state = self._get_zing_tx_state()
+            uuid = obj.getUUID()
+            tx_state.need_deletion_fact.pop(uuid, None)
+            if idxs:
+                # set this to debug
+                log.debug("buffering object update for {}".format(uuid))
+                if "path" in idxs:
+                    device_fact = ZFact.organizer_fact_from_device(obj)
+                    tx_state.need_organizers_fact[uuid] = device_fact
+                    # we also need to generate organizers facts for all the device components
+                    for comp_brain in obj.componentSearch(query={}):
+                        if not comp_brain.getUUID:
+                            continue
+                        comp_fact = ZFact.organizer_fact_from_device_component(device_fact, comp_brain.getUUID, comp_brain.meta_type)
+                        tx_state.need_organizers_fact[comp_brain.getUUID] = comp_fact
+                if "name" in idxs or "productionState" in idxs:
+                    tx_state.need_device_info_fact[uuid] = ZFact.device_info_fact(obj)
+
+    def _delete_object(self, obj):
+        if self.is_object_relevant(obj):
+            uuid = obj.getUUID()
+            log.debug("buffering object deletion for {}".format(uuid))
+            tx_state = self._get_zing_tx_state()
+            tx_state.need_deletion_fact[uuid] = ZFact.deletion_fact(uuid)
+
+    def update_object(self, obj, idxs=None):
+        """
+        ModelCatalog calls this method when an object needs to be updated
+        """
+        try:
+            self._update_object(obj, idxs)
+        except Exception as ex:
+            log.exception(traceback.format_exc())
```

Comment:
This will double-log the exception. `log.exception("message")` is the same as `log.error("%s\nmessage", ex)`. Just do the former.